GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m

```diff
@@ -330,21 +330,100 @@ - (instancetype)init {
   return self;
 }
 
+/** Creates a NSMutableDictionary for mapping bgID to task semaphore.
+ *
+ * @return A NSMutableDictionary singleton instance for mapping bgID to task semaphore.
+ */
++ (NSMutableDictionary<NSNumber *, dispatch_semaphore_t> *)getBgIDToSemaphoreMap {
+  static NSMutableDictionary<NSNumber *, dispatch_semaphore_t> *sBgIDToSemephoreMap;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    sBgIDToSemephoreMap = [[NSMutableDictionary alloc] init];
+  });
+  return sBgIDToSemephoreMap;
+}
+
+/** Creates a queue on which all task semaphore will populate and remove from map.
+ *
+ * @return A dispatch queue singleton instance for the lifetime of an application.
+ */
++ (dispatch_queue_t)getSemaphoreQueue {
```

Comment:
Here I separate the queue init and dic init, because we need to manipulate the dic on the queue and the queue should init before dic.

Reply:
things that occur within a dispatch_once are thread-safe. If the queue and dictionary are initialized at the same time and before a get or set on the dictionary is called (where the queue is needed), then it should be fine. The difficulty is generally determining where "before a get or set on the dictionary is called" is. It could be in +load, +initialize, or using a shared dispatch_once_t with two dispatch_once calls each in the get and set methods.

Reply:
Moved initialization of queue and dictionary to +(void)initialize in GDTCORApplication.